lib/include/jxl/encode.h

```diff
@@ -361,14 +363,64 @@ JXL_EXPORT JxlEncoderStatus JxlEncoderProcessOutput(JxlEncoder* enc,
+ * @param frame_options set of options and metadata for this frame. Also
+ * includes reference to the encoder object.
  * @param buffer bytes to read JPEG from. Owned by the caller and its contents
  * are copied internally.
  * @param size size of buffer in bytes.
  * @return JXL_ENC_SUCCESS on success, JXL_ENC_ERROR on error
  */
 JXL_EXPORT JxlEncoderStatus JxlEncoderAddJPEGFrame(
-    const JxlEncoderOptions* options, const uint8_t* buffer, size_t size);
+    const JxlEncoderOptions* frame_options, const uint8_t* buffer, size_t size);
+
+/**
+ * Sets the frame header with animation information for this frame.
+ *
+ * It is not required to use this function, however if have_animation is set
+ * to true in the basic info, then this function should be used to set the
+ * time duration of this individual frame. By default individual frames have a
+ * time duration of 0, making them form a composite still. See @ref
+ * JxlFrameHeader for more information.
+ *
+ * This information is stored in the JxlEncoderOptions and so is used for any
+ * frame encoded with these JxlEncoderOptions. It is ok to change between @ref
+ * JxlEncoderAddImageFrame calls, each added image frame will have the frame
+ * header that was set in the options at the time of calling
+ * JxlEncoderAddImageFrame.
+ *
+ * The is_last and name_length fields of the JxlFrameHeader are ignored, use
+ * @ref JxlEncoderCloseFrames to indicate last frame, and @ref
+ * JxlEncoderSetFrameName to indicate the name and its length instead. Calling
+ * this function will clear any name that was previously set with @ref
+ * JxlEncoderSetFrameName.
+ *
+ * @param frame_options set of options and metadata for this frame. Also
+ * includes reference to the encoder object.
+ * @param frame_header frame header data to set.
+ * @return JXL_ENC_SUCCESS on success, JXL_ENC_ERROR on error
+ */
+JXL_EXPORT JxlEncoderStatus JxlEncoderSetFrameHeader(
+    const JxlEncoderOptions* frame_options, const JxlFrameHeader* frame_header);
+
+/**
+ * Sets the name of the animation frame. This function is optional, frames are
+ * not required to have a name. This setting is a part of the frame header, and
+ * the same principles as for @ref JxlEncoderSetFrameHeader apply. The
+ * name_length field of JxlFrameHeader is ignored by the encoder, this function
+ * determines the name length instead.
+ *
+ * Calling @ref JxlEncoderSetFrameHeader clears any name that was previously
+ * set.
+ *
+ * @param frame_options set of options and metadata for this frame. Also
+ * includes reference to the encoder object.
+ * @param frame_name name of the frame to set.
+ * @param name_length length of the frame name, excluding null-termination char.
+ * @return JXL_ENC_SUCCESS on success, JXL_ENC_ERROR on error
+ */
+JXL_EXPORT JxlEncoderStatus
+JxlEncoderSetFrameName(const JxlEncoderOptions* frame_options,
+                       const char* frame_name, size_t name_length);
```

Comment:
This allows frame names to contain null chars, which might be useful and is technically allowed by the spec, but it might be better/easier to just use a C string without explicit length, so it cannot have nulls in the middle. The spec also says that the name is to be interpreted as a UTF-8 string, which kind of implies that a null byte terminates it since UTF-8 does not have multi-byte sequences containing null bytes, allowing them to be C strings.

In any case it would be good to add to the documentation that the maximum length (in bytes) of the frame_name (excluding null at the end) is 1071 bytes (largest size that can be represented is 48 + u(10)).

Reply:
Added the comment about 1071 bytes

Also removed the size parameter. For consistency we should also remove it from JxlEncoderSetExtraChannelName (in a different MR, added a TODO in this one)

I actually see advantages either way, giving the size has as advantage that you can leave out the null termination character which can be needed if you are getting the string from some encoded format in memory where it's not present. But for this MR I'll do it without name_length, a later MR that makes JxlEncoderSetFrameName and JxlEncoderSetExtraChannelName consistent can discuss this issue instead